### PR TITLE
Automox 2.0.0: Fixes to the Vulnerability Sync Actions

### PR DESCRIPTION
**Pull request by ax-tschmidtke**
### Description

  - These updates provide support for the new Vulnerability Sync APIs. Changes were made to these APIs recently that are not compatible with the existing plugin actions.
  - Makes the return values explicit for the `update_device`, `delete_device` `run_device_command`, `update_group`, and `delete_group` actions as they returned `bool` values, but were passing a dictionary return value.
  - Fixes a bug in the `update_device` action where the default parameter value for `server_group_id` would cause the API call to fail because the value would get set to `0` when not provided, instead of the current value for the device.

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [x] Screenshot of job output with the plugin changes
- [x] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [x] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [x] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [x] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [x] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [x] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [x] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [x] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [x] Work fully completed
- [x] Functional
  - [] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

#### Workflow

Screenshot 2023-10-13 at 12 22 33 AM
Screenshot 2023-10-13 at 12 22 44 AM

#### Job Results

Screenshot 2023-10-13 at 12 21 11 AM
Screenshot 2023-10-13 at 12 21 27 AM
Screenshot 2023-10-13 at 12 21 38 AM
Screenshot 2023-10-13 at 12 21 47 AM
Screenshot 2023-10-13 at 12 22 02 AM
Screenshot 2023-10-13 at 12 22 09 AM
Screenshot 2023-10-13 at 12 22 14 AM

#### Device/Group Action Workflow

Screenshot 2023-10-13 at 1 31 56 AM
Screenshot 2023-10-13 at 1 32 08 AM

#### Device/Group Action Job Results

Screenshot 2023-10-13 at 1 31 16 AM
Screenshot 2023-10-13 at 1 31 28 AM
Screenshot 2023-10-13 at 1 31 36 AM
Screenshot 2023-10-13 at 1 31 42 AM

### CODE REVIEW

**Comment by cmcnally-r7** [comment]
> Hi InsightConnect Plugins team, any idea what the ETA for review might look like? We have customers that are impacted by the bugs that this update fixes. I'm hoping to get this in their hands soon :)

Hi @ax-tschmidtke - Planning to review today and get additional reviews, latest, by the end of the week 😃

File: plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import insightconnect_plugin_runtime
+from .schema import ExecuteVulnerabilitySyncActionsInput, ExecuteVulnerabilitySyncActionsOutput, Input, Output, \
+    Component
```
Comment (cmcnally-r7) [inline plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py:3]:
```suggestion
from .schema import ExecuteVulnerabilitySyncActionsInput, ExecuteVulnerabilitySyncActionsOutput, Input, Output, Component
```

File: plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    def run(self, params={}):
+        org_id = params.get(Input.ORG_ID)
+        action_set_id = params.get(Input.ACTION_SET_ID)
+        actions = params.get(Input.ACTIONS)
+        if actions is None or len(actions) == 0:
+            raise insightconnect_plugin_runtime.exceptions.PluginException(
+                cause="Invalid input",
+                assistance="Actions cannot be empty."
+            )
```
Comment (cmcnally-r7) [inline plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py:26]: If actions is marked as `required: true` in the `plugin.spec.yaml`, then will this code block ever be hit?

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]
Just a couple of small changes 😄 Kindly can you change the branch you are merging to from `master` to `develop`

File: plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py (hunk 2 above)
Comment (ax-tschmidtke) [reply plugins/automox/icon_automox/actions/execute_vulnerability_sync_actions/action.py:26]: In my testing I found that an empty array could be sent even though actions was marked as required, so I added this to handle that situation

**Comment by ax-tschmidtke** [comment]
@cmcnally-r7 thank you for the review! I've updated the merge destination to develop and addressed your feedback.

**Review by cmcnally-r7 — APPROVED** [review]

File: plugins/automox/icon_automox/actions/list_vulnerability_sync_action_sets/action.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            "source_type:in[]": params.get(Input.SOURCE_TYPE_IN),
+            "status:in[]": params.get(Input.STATUS_IN),
+            "status:not_in[]": params.get(Input.STATUS_NOT_IN),
+            "configuration_id:equals": params.get(Input.CONFIGURATION_ID_EQUALS),
+            "configuration_id:is_set": params.get(Input.CONFIGURATION_ID_IS_SET),
+            "include_all_runs:equals": params.get(Input.INCLUDE_ALL_RUNS_EQUALS),
+        })
+        if group_by is not None:
+            params["groupBy"] = "configuration_id"
+        print(params)
```
Comment (ablakley-r7) [inline plugins/automox/icon_automox/actions/list_vulnerability_sync_action_sets/action.py:33]: Was this print statement left in on purpose?

File: plugins/automox/plugin.spec.yaml [hunk 4]
```diff
@@ -598,180 +598,252 @@ types:
...
+          status: "in_progress"
+          deleted: false
+          ip_addrs_private:
+            - 10.0.0.1
+    vulnerabilities:
+      title: Vulnerabilities
+      description: List of vulnerabilities associated with the solution
+      type: "[]solution_vulnerability"
+      required: false
+      example:
```
Comment (ablakley-r7) [inline plugins/automox/plugin.spec.yaml:838]: I don't believe these examples will be used. They shouldn't upset anything if left in however.

File: plugins/automox/help.md [hunk 5]
```diff
@@ -980,206 +682,131 @@ Example input:
...
+##### Output
 
-This action is used to update Automox device.
+|Name|Type|Required|Description|
+|----|----|--------|-----------|
+|device|device|False|The matched Automox device|
+
+Example output:
+
+```
```
Comment (ablakley-r7) [inline plugins/automox/help.md:760]: Can example outputs please be added?

File: plugins/automox/plugin.spec.yaml [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
+products: [ insightconnect ]
 name: automox
 title: Automox
 description: Automox is modernizing IT operations with continuous visibility, insight, and agility for your entire IT environment
-version: 1.2.0
-supported_versions: ["All as of 1/21/2022"]
+version: 2.0.0
+supported_versions: [ "All as of 10/09/2023" ]
 vendor: automox
 support: partner
```
Comment (ablakley-r7) [inline plugins/automox/plugin.spec.yaml:10]: Can SDK please be added here and the version updated to the latest.

Example:
sdk:
  type: slim
  version: 5
  user: nobody

File: plugins/automox/help.md [hunk 7]
```diff
@@ -1188,50 +815,25 @@ Example input:
...
-
-
 ## Troubleshooting
 
 _This plugin does not contain any troubleshooting information._
 
 # Version History
 
-* 1.2.0 - Get device by IP and Get device by hostname: fix validation issue when IP or hostname not found | Add unit tests 
+* 2.0.0 - Fix Vulnerability Sync API Actions
```
Comment (ablakley-r7) [inline plugins/automox/help.md:831]: Is it possible to list what actions have been updated and that the SDK is also updated here please?

**Review by ablakley-r7 — CHANGES_REQUESTED** [review]
Hi, the update looks great but I have a few comments. We also have new tooling now available which includes some improvements over the old tooling.

[Insight-Plugin](https://docs.rapid7.com/insightconnect/getting-started)

File: plugins/automox/icon_automox/util/cleaner.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+def clean_ax_response(a: dict) -> dict:
+    """
+    Remove null values from the action set
+    :param a:
+    :return:
+    """
+    return dict(filter(lambda k: k[1] is not None, a.items()))
```
Comment (igorski-r7) [inline plugins/automox/icon_automox/util/cleaner.py:7]: In general single letter variables are not quite a good idea 😄

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
In general lgtm but left some comments. When adding typehints try to use a common pattern. For example in some places you use `Dict` from typing and others general `dict` type 😄

**Comment by ax-tschmidtke** [comment]
> Hi, the update looks great but I have a few comments. We also have new tooling now available which includes some improvements over the old tooling.
> 
> [Insight-Plugin](https://docs.rapid7.com/insightconnect/getting-started)

@ablakley-r7 is the tooling available somewhere else? I'm unable to find it via the homebrew tap. I do have `icon-plugin` installed.

image

**Comment by ax-tschmidtke** [comment]
> In general lgtm but left some comments. When adding typehints try to use a common pattern. For example in some places you use `Dict` from typing and others general `dict` type 😄

@igorski-r7 I've addressed this in my latest commit.

**Comment by ax-tschmidtke** [comment]
@ablakley-r7 @igorski-r7 @cmcnally-r7 thank you again for the reviews! I believe that I've addressed all of the feedback with my latest commit.

**Review by ablakley-r7 — APPROVED** [review]

**Comment by ax-tschmidtke** [comment]
What does the merge/release process look like? I'd like to give our customer an update ETA for when they can expect to see the plugin available.

**Review by igorski-r7 — APPROVED** [review]

**Comment by cmcnally-r7** [comment]
> What does the merge/release process look like? I'd like to give our customer an update ETA for when they can expect to see the plugin available.

@ax-tschmidtke 
I will merge this now and if the build is successful I will release it to production tomorrow morning 😄

**Comment by ax-tschmidtke** [comment]
Awesome, thank you so much everyone!

**Pull request merged**